core/src/main/java/org/elasticsearch/rest/action/support/RestActions.java

```diff
@@ -100,7 +109,16 @@ public static QuerySourceBuilder parseQuerySource(RestRequest request) {
         if (defaultOperator != null) {
             queryBuilder.defaultOperator(Operator.fromString(defaultOperator));
         }
-        return new QuerySourceBuilder().setQuery(queryBuilder);
+        return queryBuilder;
+    }
+
+    public static SearchSourceBuilder getRestSearchSource(BytesReference sourceBytes, IndicesQueriesRegistry queryRegistry)
+            throws IOException {
+        XContentParser parser = XContentFactory.xContent(sourceBytes).createParser(sourceBytes);
+        QueryParseContext queryParseContext = new QueryParseContext(queryRegistry);
```

Comment:
wherever we create or reset the `QueryParseContext` manually, I am afraid we end up not setting the `ParseFieldMatcher`, and its default is `EMPTY` in that case. The BaseRestHandler holds the parseFieldMatcher created depending on settings, I think we should set it. It is very error-prone though, it is too easy to forget about it, maybe we should change the api a bit, for instance pass it in with the reset or as a constructor argument?